### ISSUE TITLE
Running gas explicitly returns errors not included in drone builds

### ISSUE THREAD

**Issue by zjs** [vmware/vic#6852]
Drone builds, such as [the most recent successful build on `master`](https://ci.vcna.io/vmware/vic/15115) include output like the following, which suggests that `gas` did not identify any issues:

```
...
checking whitespace...
building /go/bin/gas...
checking security problems
building /go/bin/swagger
...
```

However, running `make gas` from my machine does identify issues:

```
~/go/src/github.com/vmware/vic (master)$ make gas
Generating dependency set for cmd/imagec/
Generating dependency set for cmd/vic-machine/
Generating dependency set for cmd/vicadmin/
Generating dependency set for cmd/port-layer-server/
Generating dependency set for cmd/vic-dns/
Generating dependency set for cmd/tether/
Generating dependency set for cmd/vic-ui/
Generating dependency set for cmd/vic-machine-server/
Generating dependency set for cmd/docker/
Generating dependency set for cmd/vic-init/
Generating dependency set for cmd/rpctool/
Generating dependency set for cmd/gandalf/
checking security problems
Results:

[~/go/src/github.com/vmware/vic/lib/install/validate/config_to_data.go:82] - Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > setVCHResources(op, parent, d)


[~/go/src/github.com/vmware/vic/lib/install/validate/config_to_data.go:83] - Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > setApplianceResources(op, vm, d)


[~/go/src/github.com/vmware/vic/lib/vicadmin/validate.go:255] - Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > sess.Logout(ctx)


[~/go/src/github.com/vmware/vic/lib/vspc/vspc.go:222] - Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > conn.SetReadDeadline(time.Now().Add(remoteConnReadDeadline))


[~/go/src/github.com/vmware/vic/cmd/port-layer-server/main.go:48] - Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > pprof.StartPprof("portlayer server", pprof.PortlayerPort)


[~/go/src/github.com/vmware/vic/cmd/vic-dns/main.go:56] - Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > fmt.Fprintf(os.Stderr, fmt.Sprintf("%s : %s", r, debug.Stack()))


[~/go/src/github.com/vmware/vic/cmd/vic-machine-server/main.go:61] - Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > parser.AddGroup("Profiling Options", "", &pprofFlags)


[~/go/src/github.com/vmware/vic/pkg/logmgr/logmgr.go:242] - Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > logrotateLock.Release()


[~/go/src/github.com/vmware/vic/pkg/vsphere/tags/rest_client.go:97] - Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > c.Login(ctx)


Summary:
   Files: 432
   Lines: 85698
   Nosec: 190
  Issues: 9

Makefile:247: recipe for target 'gas' failed
make: *** [gas] Error 1
```

It does not seem like my local output should differ from drone's.

(*Note*: This issue is _not_ about the specific gas issues; it's about ensuring that the build is correctly reporting whatever issues are present.)

**Comment by mdubya66** [comment]
@anchal-agrawal touched this with https://github.com/vmware/vic/pull/4587. maybe he has insight.

**Comment by mdubya66** [comment]
@anchal-agrawal any input on this?

**Comment by zjs** [comment]
We **must** ensure gas is enforced by the build for 1.3.0.

On master, we should actually fix the errors. On the release branch, we can chose between fixing and ignoring each of the failures on a case-by-case basis (considering the complexity of each fix).

**Comment by anchal-agrawal** [comment]
Local `gas` output should not differ from drone's, however, we've seen that happen previously (cc @dougm @matthewavery). I'm not sure of what env differences can contribute to the varying behavior - with #4587 I enforced a strict `gas` check and attempted to address all issues reported by it. `gas` was also vendor'd to the latest commit.

Notably, all warnings above have `LOW` severity - while these should be addressed as well, I'm not sure why these don't appear in drone.

**Comment by jzt** [comment]
@zjs Out of curiosity, what version of go are you running?

Not sure it matters, but I'm running `make gas` directly in the CI container and not seeing any issues. I get tons of problems when I run it on my local machine using a different go version (1.9.2).

**Comment by mlh78750** [comment]
So is this not a bug and instead an issue with @zjs 's local build env?

**Comment by zjs** [comment]
> Update: I just successfully ran make gas in golang:1.8.5 and golang:1.9.2 containers with no errors.

I agree. When I run `make gas` within a container (using 1.8.5 or 1.9.2) I get no errors.

However, when I run `make gas` outside of a container (using 1.8.5 or 1.9.2), I do get errors. The errors I get seem to be valid issues.

For me, that's the crux of this issue: there are valid gas issues that are not reported the way we run gas in the build (i.e., in a container).

**Comment by anchal-agrawal** [comment]
Summary from my local experiments below. Note: all `make gas` command were run on the `releases/1.3.0` branch in the `gcr.io/eminent-nation-87317/vic-integration-test:1.38` container (the image used in CI) like so:
```
$ sudo docker run -v $(pwd):/go/src/github.com/vmware/vic -w /go/src/github.com/vmware/vic gcr.io/eminent-nation-87317/vic-integration-test:1.38 make gas 
```

Current `releases/1.3.0` branch:
```
$ sudo docker run -v $(pwd):/go/src/github.com/vmware/vic -w /go/src/github.com/vmware/vic gcr.io/eminent-nation-87317/vic-integration-test:1.38 make gas
...
building /go/bin/gas...
checking security problems
(no issues reported)
```

I made a few changes to see how `gas` responds:

```diff
diff --git a/lib/install/management/appliance.go b/lib/install/management/appliance.go
index 5e6c0d8..65cc1dc 100644
--- a/lib/install/management/appliance.go
+++ b/lib/install/management/appliance.go
@@ -851,7 +851,7 @@ func (d *Dispatcher) CheckDockerAPI(conf *config.VirtualContainerHostConfigSpec,
                // TLS enabled
                proto = "https"

-               // #nosec: TLS InsecureSkipVerify set true
+               // TLS InsecureSkipVerify set true
                tr := &http.Transport{
                        TLSClientConfig: &tls.Config{
                                InsecureSkipVerify: true,
@@ -904,7 +904,7 @@ func (d *Dispatcher) CheckDockerAPI(conf *config.VirtualContainerHostConfigSpec,
                                addr, err := addrToUse(d.op, []net.IP{cip}, cert, conf.CertificateAuthorities)
                                if err != nil {
                                        d.op.Debug("Unable to determine address to use with remote certificate, checking SANs")
-                                       // #nosec: Errors unhandled .
+                                       // Errors unhandled.
                                        addr, _ = viableHostAddress(d.op, []net.IP{cip}, cert, conf.CertificateAuthorities)
                                        d.op.Debugf("Using host address: %s", addr)
                                }
diff --git a/lib/vicadmin/validate.go b/lib/vicadmin/validate.go
index 39fd625..3232f7e 100644
--- a/lib/vicadmin/validate.go
+++ b/lib/vicadmin/validate.go
@@ -151,7 +151,7 @@ func NewValidator(ctx context.Context, vch *config.VirtualContainerHostConfigSpe
                } else {
                        v.FirewallStatus = BadStatus
                        for _, err := range firewallIssues {
-                               // #nosec: this method will not auto-escape HTML. Verify data is well formed.
+                               // this method will not auto-escape HTML. Verify data is well formed.
                                v.FirewallIssues = template.HTML(fmt.Sprintf("%s<span class=\"error-message\">%s</span>\n", v.FirewallIssues, err))
                        }
                }
diff --git a/pkg/ip/ip.go b/pkg/ip/ip.go
index 41da081..5473921 100644
--- a/pkg/ip/ip.go
+++ b/pkg/ip/ip.go
@@ -97,7 +97,7 @@ func (i *Range) Network() *net.IPNet {
 func ParseRange(r string) *Range {
        var first, last net.IP
        // check if its a CIDR
-       // #nosec: Errors unhandled
+       // Errors unhandled
        _, ipnet, _ := net.ParseCIDR(r)
        if ipnet != nil {
                first = ipnet.IP
```

With the above diff applied:
```
$ sudo docker run -v $(pwd):/go/src/github.com/vmware/vic -w /go/src/github.com/vmware/vic gcr.io/eminent-nation-87317/vic-integration-test:1.38 make gas
...
building /go/bin/gas...
checking security problems
Results:

[/go/src/github.com/vmware/vic/lib/install/management/appliance.go:857] - TLS InsecureSkipVerify set true. (Confidence: HIGH, Severity: HIGH)
  > InsecureSkipVerify: true


[/go/src/github.com/vmware/vic/lib/vicadmin/validate.go:155] - this method will not auto-escape HTML. Verify data is well formed. (Confidence: LOW, Severity: MEDIUM)
  > template.HTML(fmt.Sprintf("%s<span class=\"error-message\">%s</span>\n", v.FirewallIssues, err))


[/go/src/github.com/vmware/vic/pkg/ip/ip.go:101] - Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > _, ipnet, _ := net.ParseCIDR(r)


Summary:
   Files: 431
   Lines: 85673
   Nosec: 186
  Issues: 3

make: *** [gas] Error 1
Makefile:247: recipe for target 'gas' failed
```

In the above output, we can see that the following change (1) didn't cause a warning, though it should have:
```diff
--- a/lib/install/management/appliance.go
+++ b/lib/install/management/appliance.go
...
@@ -904,7 +904,7 @@ func (d *Dispatcher) CheckDockerAPI(conf *config.VirtualContainerHostConfigSpec,
                                addr, err := addrToUse(d.op, []net.IP{cip}, cert, conf.CertificateAuthorities)
                                if err != nil {
                                        d.op.Debug("Unable to determine address to use with remote certificate, checking SANs")
-                                       // #nosec: Errors unhandled .
+                                       // Errors unhandled.
                                        addr, _ = viableHostAddress(d.op, []net.IP{cip}, cert, conf.CertificateAuthorities)
```

while a similar change (2) did cause a warning:
```diff
--- a/pkg/ip/ip.go
+++ b/pkg/ip/ip.go
@@ -97,7 +97,7 @@ func (i *Range) Network() *net.IPNet {
 func ParseRange(r string) *Range {
        var first, last net.IP
        // check if its a CIDR
-       // #nosec: Errors unhandled
+       // Errors unhandled
        _, ipnet, _ := net.ParseCIDR(r)
```

Things to note:
* I made two changes to `lib/install/management/appliance.go`, one of which was reported and one wasn't. This _should_ rule out the possibility that this file was completely skipped by the gas tool while processing.
* Regarding (1) and (2), the function return signatures are similar in that they both return multiple values, one of which is an error. It's not clear one prompted a warning and the other didn't.
* The makefile runs the gas tool as `/go/bin/gas -skip=*_responses.go -quiet lib/... cmd/... pkg/... 2> /dev/null` which throws away stderr (which includes errors while processing files and files that were skipped). There was a change in the gas tool where if there is a fatal error during processing, it will halt (see https://github.com/GoASTScanner/gas/pull/112 and https://github.com/GoASTScanner/gas/issues/111). Our vendor'd gas tool has this commit, and I didn't see gas's execution being halted for our code.

In summary, I'm not sure why @zjs's env is producing different results. It does seem that the gas tool is flaky for the `Errors unhandled` warnings, as seen in my experiment and @zjs's output). The only possible variables are the go version and other local env differences, and running in the CI container image should address both. I didn't see any reported warnings from the gas tool for the `releases/1.3.0` branch in this scenario.

**Comment by anchal-agrawal** [comment]
Taking out of the 1.3 release and setting medium priority. I'll leave this in Not Ready (pending an estimate) and leave myself as assignee.

cc @mlh78750

**Comment by zjs** [comment]
We should see if we still observe this issue after #8277 is submitted. If so, we should consider upgrading from `gas` to `gosec`; that would be a valuable investment regardless, and may help with this issue :crossed_fingers:.

**Comment by renmaosheng** [comment]
Hi Zach,
     Given that you already have a patch, I would expect you will merge it for 1.5.1, please let me know if this is not the case, thanks.
